libnestutil/propagator_stability.h

```diff
@@ -23,9 +23,154 @@
 #ifndef PROPAGATOR_STABILITY_H
 #define PROPAGATOR_STABILITY_H
 
-// Propagators to handle similar tau_m and tau_syn_* time constants.
-// For details, please see doc/userdoc/model_details/IAF_neurons_singularity.ipynb.
-double propagator_31( double tau_syn, double tau, double C, double h );
-double propagator_32( double tau_syn, double tau, double C, double h );
+// C++ includes:
+#include <cmath>
+#include <tuple>
+
+// Includes from libnestutil:
+#include "numerics.h"
+
+
+/**
+ * Propagator class for handling similar tau_m and tau_syn_* time constants.
+ *
+ * Constants are calculated in the constructor, while
+ * propagator_31 and propagator_32 are calculated in `evaluate( h )`.
+ *
+ * For details, please see doc/userdoc/model_details/IAF_neurons_singularity.ipynb.
+ */
+class Propagator
+{
+public:
+  /**
+   * Empty constructor needed for initialization of buffers.
+   */
+  Propagator();
+
+  /**
+   * @param tau_syn Time constant of synaptic current in ms
+   * @param tau_m Membrane time constant in ms
+   * @param c_m Membrane capacitance in pF
+   */
+  Propagator( double tau_syn, double tau_m, double c_m );
+
+protected:
+  /**
+   * Calculate propagator 32 and return value along with exponentials.
+   *
+   * Exponentials are returned so they can be used directly when calculating P31.
+   *
+   * @param h time step
+   *
+   * @returns tuple with propagator 32, exp( -h / tau_syn_ ), expm1( -h / tau_m_ + h / tau_syn_ )
+   * and exp( -h / tau_m_ ), all as doubles
+   */
+  std::tuple< double, double, double, double > evaluate_P32_( double h ) const;
+
+  double tau_syn_; //!< Time constant of synaptic current in ms
+  double tau_m_;   //!< Membrane time constant in ms
+  double c_m_;     //!< Membrane capacitance in pF
+
+  double alpha_; //!< 1/(c*tau*tau) * (tau_syn - tau)
+  double beta_;  //!< tau_syn * tau/(tau - tau_syn)
+  double gamma_; //!< beta_/c
```

Comment:
Since propagator parameters cannot change once objects are instantiated, maybe making them const can allow for further optimization.
```suggestion
  const double tau_syn_; //!< Time constant of synaptic current in ms
  const double tau_m_;   //!< Membrane time constant in ms
  const double c_m_;     //!< Membrane capacitance in pF

  const double alpha_; //!< 1/(c*tau*tau) * (tau_syn - tau)
  const double beta_;  //!< tau_syn * tau/(tau - tau_syn)
  const double gamma_; //!< beta_/c
```

Reply:
@heplesser If I make these `const` I get the following error message, one per variable: `error: non-static const member ‘const double Propagator::tau_syn_’, can’t use default assignment operator`. I am not quite sure how to fix this though.

 `‘Propagator& Propagator::operator=(Propagator&&)` is also implicitly deleted, so it is no longer possible to do `propagator_ex_ = PropagatorAlpha( P_.tau_syn_ex_, P_.tau_m_, P_.c_m_ );`

Reply:
Given that we have and `update_constants()` method, making the members `const` seems contradictory, but then so seems the idea of updating *constants*. I took another look at the code and got some ideas, which I will try out and then get back.

Reply:
@stinebuu Please excuse my previous nonsensical comment—I have looked at an outdated version of the code due to some Git confusion. I think we just need to leave things as they are, i.e., non-const.